feat: Mass Minting in drop page

hassnian opened the pull request:
## PR Type

- [ ] Bugfix
- [x] Feature
- [x] Refactoring


## Context

- [x] Closes #9136
- [x] Closes https://github.com/kodadot/private-workers/issues/84

## Needs QA check

- @kodadot/qa-guild please review

- amount of nfts that you can mint depends on:
    - paid: only (balance / price)
    - holderOf: min between (balance / price) and amount of available nfts 

- when minting one nft , ensure it's not generating a new one , ie it's using the same 

![CleanShot 2024-03-04 at 13 26 32@2x](https://github.com/kodadot/nft-gallery/assets/44554284/38bf972b-2209-452c-aee7-5945683a6bc1)

- when multiple nfts are being generated , no duplicates should be generated
- **paid , paid + holderOf , holdeof** , mass mint code has been added  , **free** code refactoring but still should be tested 

## Needs Design check

- @exezbcz please review

#### Did your issue had any of the "$" label on it?

- [x] Fill up your DOT address: [Payout](https://kodadot.xyz/dot/transfer?target=16faLfsywwNATaEfbH2ah75dn6ZmctQWpMS5G4KFhbmj5hnD)


## Screenshot 📸

- [ ] My fix has changed **something** on UI;

yangwao commented:
Will we make it to release? That would be amazing if so

hassnian commented:
that was my aim from the get-go, will try , but I doubt it 

I'd want @prury to be able to test it well since a lot of drop page code is changing.

also found out this https://github.com/kodadot/private-workers/issues/84 is not implemented and it's needed.

yangwao commented:
Okay 👌

Let's aim for the next release

yangwao commented:
Time to test it?

hassnian commented:
Once we have https://github.com/kodadot/private-workers/pull/86 merged

hassnian commented:
@prury  this is important 
![CleanShot 2024-03-03 at 14 37 29@2x](https://github.com/kodadot/nft-gallery/assets/44554284/68f7ab64-b177-4c3c-87f1-8689e5b54356)

otherwise each time you  massmint you'll be allocated new sn numbers

also all drop types have been received changes , free, paid , HolderOf, so once we have the worker ready please test them 

thanks

hassnian commented:
if you see the same sn , like this 
![CleanShot 2024-03-04 at 16 53 18@2x](https://github.com/kodadot/nft-gallery/assets/44554284/747bb15c-5969-4b46-b80c-e3bda79a1ed7)

it's because the worker is not deployed

prury commented:
buttons keeps loading even if i did not click:
https://github.com/kodadot/nft-gallery/assets/36627808/c685a0eb-5914-49d4-bb1b-45a0a416023c

still not ready for testing i guess?
 
https://github.com/kodadot/nft-gallery/assets/36627808/44d252b2-409a-4afb-a196-396ac14dcc16

hassnian commented:
> buttons keeps loading even if i did not click: https://github.com/kodadot/nft-gallery/assets/36627808/c685a0eb-5914-49d4-bb1b-45a0a416023c
> 

checking

> still not ready for testing i guess? https://github.com/kodadot/nft-gallery/assets/36627808/44d252b2-409a-4afb-a196-396ac14dcc16
> 
> second try: https://github.com/kodadot/nft-gallery/assets/36627808/6d1c807d-e785-4400-8990-a71173f269f3

yep , the worker is not deployed cc @preschian

preschian commented:
@prury @hassnian I just deployed the worker, can check it again 🙏🏻

preschian commented on composables/drop/useDropMassMint.ts:
```diff
@@ -0,0 +1,327 @@
+import {
+  AllocateCollectionResult,
```
missing?

hassnian replied:
yup, mb missed it , fixed

preschian commented on components/collection/drop/GenerativeLayout.vue:
```diff
@@ -102,10 +106,13 @@ import type {
 const props = withDefaults(
   defineProps<{
     collectionId: string
+    amountToMint?: number
```
eslint "requires default value to be set"

hassnian replied:
fixed

preschian reviewed (approved):
otherwise lgtm

hassnian commented:
> > buttons keeps loading even if i did not click: https://github.com/kodadot/nft-gallery/assets/36627808/c685a0eb-5914-49d4-bb1b-45a0a416023c
> 
> checking
> 

can't reproduce , it's still happening?. drop name ?

exezbcz commented:
i cant set the number when i am not connected? 


https://github.com/kodadot/nft-gallery/assets/90852205/e3a0f8ff-f930-44ef-b63a-8ad7d423db02

is it feature?

hassnian commented:
> i cant set the number when i am not connected?
> 
>  Screen.Recording.2024-03-05.at.12.18.48.mov 
> is it feature?
> 
> actually does not work even when connected
> 
>  Screen.Recording.2024-03-05.at.12.20.00.mov

had a chat with @exezbcz no balance checking should be done for paid drops , the user can add more funds inside the modal and be able to mint , so no max mint amount for paid drops, fixing

hassnian commented:
> > > buttons keeps loading even if i did not click: https://github.com/kodadot/nft-gallery/assets/36627808/c685a0eb-5914-49d4-bb1b-45a0a416023c
> > 
> > 
> > checking
> 
> can't reproduce , still happening?. drop name ?

@prury found the issue , fixing it

prury commented:
asked for everyone's review since big

prury commented:
issues found: 
- preview generation still won't stop
- thumbnails not appearing on success modal
- copy button leads to only one of them(i suppose best its to lead to person profile?)
- listing does not show numbering(not sure if from this PR)
- if listing is cancelled nfts stays on blue square selection state
- thumbnails in latest nft mints took a while to be updated 5+ minutes

https://github.com/kodadot/nft-gallery/assets/36627808/80f4c6ab-f0a7-4ca4-b6c2-7fbe84d1bdbc

daiagi commented:
@hassnian 
I think this PR should wait until https://github.com/kodadot/nft-gallery/issues/9575#issuecomment-1980156973
are done
then it should be re-written against https://github.com/kodadot/nft-gallery/tree/drops-refactor branch

hassnian commented:
> 
> * preview generation still won't stop
> * thumbnails not appearing on success modal

fixed

> * copy button leads to only one of them(i suppose best its to lead to person profile?)

had a chat with @exezbcz regarding this when I worked on #8978 , we decided that the copy url button should be removed in the case of multiple nfts, 

fixed

> * listing does not show numbering(not sure if from this PR)

nope not from this pr,  please raise a new issue , seems  related to https://github.com/kodadot/nft-gallery/issues/9144

> * if listing is cancelled nfts stays on blue square selection state

fixed

> * thumbnails in latest nft mints took a while to be updated 5+ minutes

iirc the metadata updating process is queued, not sure cc @vikiival

hassnian commented:
> @hassnian I think this PR should wait until [#9575 (comment)](https://github.com/kodadot/nft-gallery/issues/9575#issuecomment-1980156973) are done then it should be re-written against https://github.com/kodadot/nft-gallery/tree/drops-refactor branch

@daiagi  got no issues with that,  feel free to add the s-blocked tag , will come back to this pr once those issues are done

yangwao commented:
Good to put in hold to prevent some bad things from happening, but it's been a month since issue was raised, let's aim please for next release or we can do post-release if everything will be good to go

vikiival commented:
> iirc the metadata updating process is queued, not sure cc @vikiival

You can use batch uploading. 

So you batch upload all images and then batch upload all metadatas. 

Same is used by mass mint. 
@daiagi can help with this one

yangwao commented:
I guess once this is merged
- https://github.com/kodadot/nft-gallery/pull/9620
we can push this one as well?

vikiival commented:
> we can push this one as well?

Mass upload of metadata is not implemented, see the message above

hassnian commented:
> > we can push this one as well?
> 
> Mass upload of metadata is not implemented, see the message above

on it

yangwao commented:
- https://github.com/kodadot/nft-gallery/pull/9620
Close bein done, can be adjusted towards that branch and aim for next week release 🙏

hassnian commented:
@yangwao  the target of that pr is not main 

![CleanShot 2024-03-11 at 08 20 07@2x](https://github.com/kodadot/nft-gallery/assets/44554284/123f570f-4372-470d-ba00-460a69d200f0)

![CleanShot 2024-03-11 at 08 13 09@2x](https://github.com/kodadot/nft-gallery/assets/44554284/8e35fe81-6daa-48ad-8967-a69452443b99)

should I target `drops-refactor` ?

hassnian commented:
> @hassnian
> 
> ![image](https://private-user-images.githubusercontent.com/22791238/311603587-e999735f-62b5-40a7-96ed-d28d0cdf7a4a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAxMzgzMTcsIm5iZiI6MTcxMDEzODAxNywicGF0aCI6Ii8yMjc5MTIzOC8zMTE2MDM1ODctZTk5OTczNWYtNjJiNS00MGE3LTk2ZWQtZDI4ZDBjZGY3YTRhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzExVDA2MjAxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI4ZDk1MTMwMmE3MmE0YzUxYWMxMzkxODAzN2YyYTc2ZDI3YmFmMzBkZjE3YzZjMjQ4NDBhNzUzZmQ3YTBiZGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.m12wjwuTSjx7-G6250YDMSUbqRPHweKROhABK91cjko)
> 
> PR against the refactor branch please



> @hassnian
> 
> ![image](https://private-user-images.githubusercontent.com/22791238/311603587-e999735f-62b5-40a7-96ed-d28d0cdf7a4a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAxMzgzMTcsIm5iZiI6MTcxMDEzODAxNywicGF0aCI6Ii8yMjc5MTIzOC8zMTE2MDM1ODctZTk5OTczNWYtNjJiNS00MGE3LTk2ZWQtZDI4ZDBjZGY3YTRhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzExVDA2MjAxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI4ZDk1MTMwMmE3MmE0YzUxYWMxMzkxODAzN2YyYTc2ZDI3YmFmMzBkZjE3YzZjMjQ4NDBhNzUzZmQ3YTBiZGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.m12wjwuTSjx7-G6250YDMSUbqRPHweKROhABK91cjko)
> 
> PR against the refactor branch please

will be branching out from drops-refactor and adding this code manually , too many conflicts.  don't want to break something while merging

daiagi commented:
close this one then?
